samples: Add MetaIRQ dispatch sample

andyross opened the pull request:
MetaIRQs are described in docs and exercised in tests, but there's no
sample explaining how they are intended to be used to perform
interrupt bottom half processing.

This simple tool spawns a set of worker threads at different
priorities (some cooperative) which process "messages" (which require
CPU time to handle) received from a fake "device" implemented with
timer interrupts.  The device hands off the events directly to a
MetaIRQ thread which is expected to parse and dispatch them to the
worker threads.

The test demonstrates that no matter the state of the system, the
MetaIRQ thread always runs synchronously when woken by the ISR and is
able to preempt all worker threads to do its job.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
samples: Add MetaIRQ dispatch sample

MetaIRQs are described in docs and exercised in tests, but there's no
sample explaining how they are intended to be used to perform
interrupt bottom half processing.

This simple tool spawns a set of worker threads at different
priorities (some cooperative) which process "messages" (which require
CPU time to handle) received from a fake "device" implemented with
timer interrupts.  The device hands off the events directly to a
MetaIRQ thread which is expected to parse and dispatch them to the
worker threads.

The test demonstrates that no matter the state of the system, the
MetaIRQ thread always runs synchronously when woken by the ISR and is
able to preempt all worker threads to do its job.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
This seems to be working well and should be good for review.  I still need to write a README

andyross commented:
Filter out ARCH_POSIX platforms (which don't work: the test here is of
thread preemption, and native_posix can't preempt by design).

Bump stack size, riscv64 was running into trouble with the newest SDK.

Put a correctness check in to avoid the divide by zero during report generation when one of the threads gets no events at all.

Drop sys_rand32_get(), which is (counter to my earlier understanding) STILL just a timer stub and not actually random (thus the divide by zero above!) and use a custom RNG in the message device.  Sigh.

ceolin commented on samples/metairq_dispatch/src/msgdev.c:
```diff
@@ -0,0 +1,104 @@
+ * inside the ISR and needs no locking for the otherwise non-atomic
+ * state.
+ */
+static u32_t rand32(void)
```
mixed feeling about it ... It is just masking another problem.

andyross replied:
I dunno, this is a sample.  It's only going to be useful to people to inspect and emulate if it works correctly.  And right now it doesn't work correctly with sys_rand32_get().  I agree that's a bug and it should be fixed, but committing a broken sample or delaying merge while we wait for a bug fix to an unrelated subsystem seems like not a great choice either.

nashif replied:
agreed

ceolin replied:
I disagree about it just be a sample, I really believe that people start applications copying from samples. I have to be very careful about our samples to have there best practices from Zephyr development and not only about the features that is being shown.

But, I don't want to pedantic with this particular pr, since my comment apply for all of them :)
The sample is correct and is addressing that feature.

ceolin reviewed (commented):
Test is ok for me, I just think is better use Zephyr's random api and keep the check to ensure that is not dividing by 0.

aescolar commented on samples/metairq_dispatch/src/main.c:
```diff
@@ -0,0 +1,239 @@
+		 * and get queued.
+		 */
+		while (k_cycle_get_32() - start < m.proc_cyc) {
+			for (volatile int i = 0; i < 100; i++) {
```
> Filter out ARCH_POSIX platforms (which don't work: the test here is of
thread preemption, and native_posix can't preempt by design).

Just a silly clarification/comment:
POSIX arch targets can preempt the SW threads. And they will if the thread takes time to execute and some interrupt comes while it is executing. But (following the infinite clock metaphor) normally SW threads take 0 time to execute, and therefore nothing tends to happen that will cause them to lose context.
In the POSIX arch, the way to make SW threads take time is to have them busy wait. So if you add a k_busy_wait(some_small_time) here, the thread will actually spend time, and will be interrupted in the middle of that k_busy_wait()

nashif replied:
indeed, this works with native_posix if we have a k_busy_wait here.

andyross commented:
Rebased, added a README that got forgotten earlier.  Fixed up the nit with the blank lines in prj.conf.

Regarding native_posix:

> indeed, this works with native_posix if we have a k_busy_wait here.

Right, but then it's not really demonstrating what it's intended to, which is that the worker threads pull down a message and then have to spend a completely arbitrary amount of non-trivial CPU time doing completely arbitrary things, during which they must remain preemptible by the MetaIRQ thread.  If you **have** to have a k_busy_wait() (or other native_posix instrumentation) in there, then IMHO the reader is likely to be confused about exactly what the requirements are for such a thread.  

And the point is there are no requirements, you can do whatever you want and things still work.

Really ISR bottom half processing isn't really something that's appropriate to native_posix, which has a special (and obviously really clever) way of representing interrupts in a realm (unix userspace processes) that doesn't actually have them.

I added a section to the README to clarify this.

andyross commented:
> And the point is there are no requirements, you can do whatever you want and things still work.

Well, except disable interrupts. You can't do that. :)

nashif commented:
> I added a section to the README to clarify this.

there is no README :)

andyross commented:
Oops, added correctly this time

andyross commented:
FWIW: don't really understand the doc failure.  All the other samples have README.rst files, yet none of them are included in another that I can find.  Is this a known thing, did I miss a step, or should I just change it to a README.txt to evade the filter?

aescolar commented on samples/metairq_dispatch/README.rst:
```diff
@@ -0,0 +1,63 @@
+preemption behavior, it does not run without modification on
+native_posix platforms.  In that emulation environment, threads will
+not be preempted except at specific instrumentation points (e.g. in
+k_busy_wait()) where they will voluntarily release the CPU.
```
> they will voluntarily release the CPU.

Just being very picky: Note that calling k_busy_wait() the thread does not really release the CPU, but halts it (without the kernel knowing about it). Meaning, if all interrupts were masked that thread could be sure it would continue running after the wait time without anything else running in between.

It is certainly true though, that interrupts cannot come at any moment in native_posix in the middle of a SW thread execution. They can come during a k_busy_wait(), or right when a pending interrupt is unmasked, or when an unmasked interrupt is raised (e.g. a SW interrupt). (That seemed to be the easiest way of building something which provided deterministic/reproducible behavior)

(I'm not necessarily asking for this sentence to be rephrased, I just wanted to clarify things)

nashif commented:
> FWIW: don't really understand the doc failure. All the other samples have README.rst files, yet none of them are included in another that I can find. Is this a known thing, did I miss a step, or should I just change it to a README.txt to evade the filter?

I assume we will get more samples like this, so...
- Move the sample under samples/scheduler
- add the changes below:

```diff
diff --git a/samples/index.rst b/samples/index.rst
index 90a6bf7b00..ce1d6f08f9 100644
--- a/samples/index.rst
+++ b/samples/index.rst
@@ -24,6 +24,7 @@ Samples and Demos
    portability/*
    posix/*
    video/*
+   scheduler/*
    smp/*

 .. comment
diff --git a/samples/scheduler/index.rst b/samples/scheduler/index.rst
new file mode 100644
index 0000000000..62fcde3da6
--- /dev/null
+++ b/samples/scheduler/index.rst
@@ -0,0 +1,10 @@
+.. _smp-samples:
+
+Various Kernel and Scheduler Samples
+####################################
+
+.. toctree::
+   :maxdepth: 1
+   :glob:
+
+   **/*
```

andrewboie commented on samples/scheduler/metairq_dispatch/src/main.c:
```diff
@@ -0,0 +1,239 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_INF);
+
+#define STACK_SIZE 2048
```
consider using
```
#define STACK_SIZE (2048 + CONFIG_TEST_EXTRA_STACKSIZE)
```
